Encrypt TOTP keys

dd32 opened the pull request:
Fixes #101

Depends upon https://github.com/WordPress/wporg-mu-plugins/pull/390
Made possible by https://github.com/WordPress/two-factor/pull/546

NOTE: if you test this on a sandbox (and then remove this), you will be unable to login with TOTP again without deactivating TOTP.

Testing Instructions:
 - Make sure you're running a version of `two-factor` with https://github.com/WordPress/two-factor/pull/546 merged
 - Define `WPORG_ENCRYPTION_KEY` and optionally `WPORG_TWO_FACTOR_ENCRYPTION_KEY`, with a value from `\WordPressdotorg\MU_Plugins\Encryption\generate_encryption_key()`
 - View `_two_factor_totp_key` value. Validate that it appears plaintext.
 - Login with TOTP.
 - View `_two_factor_totp_key` value. Validate that it appears to be encrypted (Long, and prefixed with `$t1$`).
 - Deactivate TOTP.
 - Setup TOTP.
 - Validate everything worked as expected.

TODO:
 - [x] Unit Tests.
 - [x] https://github.com/WordPress/wporg-mu-plugins/pull/390 (This is not required prior to this merging, but is required for it to actually encrypt)
 - [x] Test once 390 is merged
 - [x] Test on production sandbox, verify TOTPs are upgraded upon login.

dd32 committed:
Add a Two_Factor_TOTP class variant that uses encryption.

dd32 committed:
Upgrade unencrypted keys to encrypted upon retrieval.

pkevan committed:
Update test to reflect new provider

dd32 commented on tests/test-wporg-two-factor.php:
```diff
@@ -229,7 +229,7 @@ public function test_set_primary_provider_for_user() {
 		$enabled = Two_Factor_Core::enable_provider_for_user( self::$regular_user->ID, 'Two_Factor_Totp' );
 
-		$expected = 'Two_Factor_Totp';
+		$expected = 'WordPressdotorg\Two_Factor\Encrypted_Totp_Provider';
 		$actual   = get_class( Two_Factor_Core::get_primary_provider_for_user( self::$regular_user->ID ) );
 		$this->assertTrue( $enabled );
 		$this->assertSame( $expected, $actual );
```
This test can probably change to both check the class, but also check it's Totp as expected, as that's what we set as active above.

```suggestion
		$this->assertTrue( $enabled );

		$provider = Two_Factor_Core::get_primary_provider_for_user( self::$regular_user->ID );

		$expected_class = 'WordPressdotorg\Two_Factor\Encrypted_Totp_Provider';
		$actual_class   = get_class( $provider );
		$this->assertSame( $expected_class, $actual_class );

		$expected_key = 'Two_Factor_Totp';
		$actual_key   = $provider->get_key();
		$this->assertSame( $expected_key, $provider->get_key() );
```

pkevan replied:
Good point, I was primarily fixing the tests before testing - makes sense to cover both.

pkevan commented:
Tested on a sandbox, and confirmed the TOTP key is encrypted, and you can log in once it has been encrypted.

dd32 committed:
Switch to using authenticated encryption. The encrypted keys are only valid per user ID.

dd32 committed:
Merge branch 'add/encrypted-totp' of github.com:WordPress/wporg-two-factor into add/encrypted-totp

dd32 committed:
Tests: Test that the class is expected, but also that it returns the expected key.

dd32 committed:
Tests: Validate that the TOTP key is stored in an encrypted format in the user_meta.

dd32 commented:
I've updated this to include the User ID as part of the encryption, such that user_meta for one user is invalid for another.
_(I've 'upgraded' the existing encrypted TOTP keys on WordPress.org)._

I've updated the unit tests to test that the key is stored in a format different than the key (ie. encrypted) but left the validation of the encryption functions to the Encryption library PR.
_edit: Of course this doesn't actually pass yet.. as the encryption methods aren't merged.. [3d89b20](https://github.com/WordPress/wporg-two-factor/pull/103/commits/3d89b204f8aa99bfb12d1e509c7479179946f126) can be reverted once the upstream PR is merged._

dd32 committed:
Tests: Only validate that it's encrypted if the encryption methods are available.

dd32 committed:
Update function names per https://github.com/WordPress/wporg-mu-plugins/pull/390

dd32 committed:
Revert "Tests: Only validate that it's encrypted if the encryption methods are available."

This reverts commit 3d89b204f8aa99bfb12d1e509c7479179946f126.

dd32 committed:
Tests: Define an encryption key for Two-factor.

dd32 committed:
Set the users TOTP key via the provider, to allow it to be encrypted.

dd32 committed:
Tests: Add a test that verifies that Secret keys are upgraded from stored non-encrypted to encrypted.
